tooling/minimize/src/constant.rs

```diff
@@ -31,6 +31,42 @@ fn translate_const_val<'cx, 'tcx>(
         Type::Tuple { fields, .. } if fields.is_empty() => {
             return ValueExpr::Tuple(List::new(), ty);
         }
+        Type::Enum { size, discriminator, variants, .. } => {
+            let Some(bits) = val.try_to_bits(rs::Size::from_bits(size.bits().try_to_u8().unwrap())) else {
+                panic!("Can only create constant enum from bits. Got {:?}", val);
+            };
+
+            // special case for bits = 0 which can be indicative of a None constant.
+            // FIXME: Allow for other constant values by using the offset and value type.
+            //        However this is probably going to require machine information for
+            //        the value encoding which we don't have here.
+            if bits == 0 {
```

Comment:
What is this doing?
This should be using the rustc interpreter APIs to get the discriminant and then recurse.
But actually what it should probably do is translate the raw bytes to a MiniRust allocation and then just load the value from there.

You don't have to worry about this for now, but I wonder why you had to write this code.

Reply:
Option<NonZeroU8>::None uses this for its constant somehow. I rewrote it to write the bytes into a global allocation, however we get the bits without endianness information (at least we know that "the first `size` bytes of `data` are the value)

Reply:
Thing is, we aren't given an allocation. We are given an `ConstValue::Scalar(Scalar::Int(ScalarInt))` where the Scalar comes from [here](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/mir/interpret/value/enum.Scalar.html) and gives us the data as a `u128`.

The ConstValue documentation claims that Scalar and Slice could be used for enum discriminant optimizations and pattern matching simplification. However the type that this constant has is the entire enum, so I think we are literally byte-comparing or instantiating the enum from those bits. Now, I have read nowhere that those bits in the u128 have any endianness, which would be required for either getting a vector of bytes to move into a global allocation (because I don't know if the u128 is in any particular endianness and whether that matches the machine's), or to translate it into a MiniRust Value because for decoding I'd once again have to know the endianness.

With this said, the smallest thing to make it work is this way of moving it into a global allocation, though as I said, getting the bytes vector is a risk because of the endianness and I have yet to encounter this with a value other than 0u128.

The smallest thing to make this work on all 0u128 in a safe manner was the previous way, to ensure that the variant that has the tag 0 has no data and return this as `Value::Variant`. This is safe because if all bits are 0 then the tag is guaranteed to be 0, but does no longer work if we don't get a 0u128 constant (except for `u128::MAX`).

Reply:
Okay, after a few more hours of digging through compiler code it seems like [constant propagation](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_transform/src/dataflow_const_prop.rs#L685) might indeed write immediates of the entire enum with the bytes in target endianness. Interestingly enough the documentation of ConstValue tells us that a constant enum represents "enum discriminants and pattern matching logic". Other than the pattern matching logic (which writes the enum bits to test into a global allocation to compare it to) I don't find anything that would write the discriminant.

Reply:
I've pushed a commit adjusting this slightly -- good enough for now. It's not the proper thing yet, but that needs a small rustc change so that we can access the right APIs: https://github.com/rust-lang/rust/pull/121396.

Reply:
Since this behaviour only occurs for `Option<NonZeroXX>` we decided to remove this from the test and open an issue for this. `consts.rs` needs some rework, see above.